docarray/index/backends/redis.py

```diff
@@ -0,0 +1,606 @@
+import uuid
+from collections import defaultdict
+from typing import (
+    TypeVar,
+    Generic,
+    Optional,
+    List,
+    Dict,
+    Any,
+    Sequence,
+    Union,
+    Generator,
+    Type,
+    cast,
+    TYPE_CHECKING,
+    Iterator,
+    Mapping,
+    Tuple,
+)
+from dataclasses import dataclass, field
+
+import json
+import numpy as np
+from numpy import ndarray
+
+from docarray.array import AnyDocArray
+from docarray.index.backends.helper import _collect_query_args
+from docarray import BaseDoc, DocList
+from docarray.index.abstract import (
+    BaseDocIndex,
+    _raise_not_composable,
+)
+from docarray.typing import NdArray
+from docarray.typing.tensor.abstract_tensor import AbstractTensor
+from docarray.utils._internal._typing import safe_issubclass
+from docarray.utils._internal.misc import import_library
+from docarray.utils.find import _FindResultBatched, _FindResult, FindResult
+
+if TYPE_CHECKING:
+    import redis
+    from redis.commands.search.query import Query
+    from redis.commands.search.field import (  # type: ignore[import]
+        NumericField,
+        TextField,
+        VectorField,
+    )
+    from redis.commands.search.indexDefinition import IndexDefinition, IndexType  # type: ignore[import]
+else:
+    redis = import_library('redis')
+
+    from redis.commands.search.field import (
+        NumericField,
+        TextField,
+        VectorField,
+    )
+    from redis.commands.search.indexDefinition import IndexDefinition, IndexType
+    from redis.commands.search.query import Query
+
+TSchema = TypeVar('TSchema', bound=BaseDoc)
+
+VALID_DISTANCES = ['L2', 'IP', 'COSINE']
+VALID_ALGORITHMS = ['FLAT', 'HNSW']
+VALID_TEXT_SCORERS = [
+    'BM25',
+    'TFIDF',
+    'TFIDF.DOCNORM',
+    'DISMAX',
+    'DOCSCORE',
+    'HAMMING',
+]
+
+
+class RedisDocumentIndex(BaseDocIndex, Generic[TSchema]):
+    def __init__(self, db_config=None, **kwargs):
+        """Initialize RedisDocumentIndex"""
+        if db_config is not None and getattr(db_config, 'index_name'):
```

Comment:
BaseDocIndex init needs `index_name` to work, that's why I assign it in advance

Reply:
where is this needed?

Reply:
what u should do is make `index_name` propery read it from `db_config`

Reply:
I still have to check the index name before the init function and create myself in case it's not provided so I don't see why your option is better

It's needed when creating subindices in `_init_subindex`